lib/SILOptimizer/Analysis/SimplifyInstruction.cpp

```diff
@@ -453,6 +454,24 @@ SILValue InstSimplifier::visitBeginAccessInst(BeginAccessInst *BAI) {
   return SILValue();
 }
 
+SILValue InstSimplifier::visitMetatypeInst(MetatypeInst *MI) {
+  auto metaType = MI->getType().castTo<MetatypeType>();
+  auto instanceType = metaType.getInstanceType();
+  // Tuple, Struct, and Enum MetatypeTypes have a single value.
+  // If this metatype is already passed as an argument reuse it to enable
+  // downstream CSE/SILCombine optimizations.
+  // Note: redundant metatype instructions are already handled by CSE.
+  if (isa<TupleType>(instanceType)
+      || instanceType.getStructOrBoundGenericStruct()
+      || instanceType.getEnumOrBoundGenericEnum()) {
```

Comment:
We could promote `hasSingletonMetatype` from `lib/SIL/TypeLowering.cpp` into a proper API and reuse it here.